// ask-sdk-controls/dist/src/logging/Logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
const tslib_1 = require("tslib");
/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
const debug_1 = tslib_1.__importDefault(require("debug"));
const DEFAULT_LOG_LEVEL = 'error:*, warn:*';
class Logger {
    constructor(moduleName) {
        var _a;
        this.moduleName = moduleName;
        const namespace = (_a = process.env.DEBUG) !== null && _a !== void 0 ? _a : DEFAULT_LOG_LEVEL;
        debug_1.default.enable(namespace);
    }
    error(message) {
        debug_1.default(`error:${this.moduleName}`)(message);
    }
    warn(message) {
        debug_1.default(`warn:${this.moduleName}`)(message);
    }
    debug(message) {
        debug_1.default(`debug:${this.moduleName}`)(message);
    }
    info(message) {
        debug_1.default(`info:${this.moduleName}`)(message);
    }
}
exports.Logger = Logger;
