// ask-sdk-controls/dist/src/constants/Strings.js
"use strict";
/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Strings = void 0;
var Strings;
(function (Strings) {
    /**
     * Built-in slot types
     */
    let SlotType;
    (function (SlotType) {
        // TODO: consolidate with AmazonBuiltInSlotType.ts
        SlotType["AmazonDate"] = "AMAZON.DATE";
        SlotType["AmazonTime"] = "AMAZON.TIME";
        SlotType["AmazonOrdinal"] = "AMAZON.Ordinal";
        SlotType["AmazonNumber"] = "AMAZON.NUMBER";
    })(SlotType = Strings.SlotType || (Strings.SlotType = {}));
    /**
     * Built-in feedback slot values
     *
     */
    let Feedback;
    (function (Feedback) {
        Feedback["Affirm"] = "affirm";
        Feedback["Disaffirm"] = "disaffirm";
    })(Feedback = Strings.Feedback || (Strings.Feedback = {}));
    /**
     * Built-in action slot values
     */
    let Action;
    (function (Action) {
        Action["Set"] = "set";
        Action["Change"] = "change";
        Action["Select"] = "select";
        Action["None"] = "none";
    })(Action = Strings.Action || (Strings.Action = {}));
    /**
     * Built-in target slot values
     */
    let Target;
    (function (Target) {
        Target["It"] = "it";
    })(Target = Strings.Target || (Strings.Target = {}));
})(Strings = exports.Strings || (exports.Strings = {}));
