// ask-sdk-controls/dist/src/utils/Predicates.js
"use strict";
/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.falseIfGuardFailed = exports.failIf = exports.okIf = exports.GuardFailed = exports.StateConsistencyError = void 0;
class StateConsistencyError extends Error {
    constructor(message) {
        super(message);
        // see: typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html
        Object.setPrototypeOf(this, new.target.prototype); // restore prototype chain
        this.name = StateConsistencyError.name; // stack traces display correctly now 
    }
}
exports.StateConsistencyError = StateConsistencyError;
/**
 * An Error class used to short-circuit a set of predicate tests
 *
 * This error indicates that a "guard" expression did not pass.
 *
 * Purpose:
 * * This helps to write predicate functions as a linear chain of individual tests.
 */
class GuardFailed extends Error {
    constructor(message) {
        super(message);
        // see: typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html
        Object.setPrototypeOf(this, new.target.prototype); // restore prototype chain
        this.name = GuardFailed.name; // stack traces display correctly now 
    }
}
exports.GuardFailed = GuardFailed;
/**
 * Allows normal code flow to proceed if predicate is true.
 *
 * If predicate is true, the code proceeds normally.
 * If false, an GuardFailed error is thrown.
 *
 * Purpose:
 * * This helps to write predicate functions as a linear chain of individual tests.
 */
function okIf(predicate) {
    if (!predicate) {
        throw new GuardFailed('okIf failed. predicate===false.');
    }
}
exports.okIf = okIf;
/**
 * Aborts normal code flow if predicate is true.
 *
 * If predicate is true, an GuardFailed error is thrown.
 * If false, the code proceeds normally.
 *
 * Purpose:
 * * This helps to write predicate functions as a linear chain of individual tests.
 * @param predicate
 */
function failIf(predicate) {
    if (predicate) {
        throw new GuardFailed('failIf triggered. predicate===true.');
    }
}
exports.failIf = failIf;
/**
 * Consumes an GuardFailed error but otherwise rethrows.
 *
 * Purpose:
 * * This helps to write predicate functions as a linear chain of individual tests.
 */
function falseIfGuardFailed(err) {
    if (err instanceof GuardFailed) {
        return false;
    }
    throw err; // otherwise rethrow
}
exports.falseIfGuardFailed = falseIfGuardFailed;
