// ask-sdk-controls/dist/src/utils/testSupport/SkillInvoker.js
"use strict";
/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the 'License').
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the 'license' file accompanying this file. This file is distributed
 * on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SkillInvoker = void 0;
const SkillWrapper_1 = require("./SkillWrapper");
/**
 * Wraps a userInterface definition as a Skill object for use in e2e scenario tests.
 * (multi-turn ASK Request -> ASK Response with state serialization)
 *
 * This wrapper mimics the lifecycle of an ASK Skill using AWS Lambda
 * that is repeatedly invoked as a 'stateless' function which receives actual state
 * in the session attributes of the Request object.
 *
 * The main part still simulated is the use of NLU & the interaction model.. we fake up the
 * IntentRequests based on ControlInput objects.
 *
 */
class SkillInvoker {
    constructor(skillOrRequestHandler) {
        this.skill = ((skillOrRequestHandler.handle !== undefined) ? SkillWrapper_1.wrapRequestHandlerAsSkill(skillOrRequestHandler) : skillOrRequestHandler);
    }
    /**
     * Invoke the skill with a control-input object.
     * The control input is first converted to a RequestEnvelope(IntentRequest)
     *
     * @param input
     * @returns [prompt ssml inner-text, reprompt ssml inner-text]
     */
    async invoke(input) {
        const envelope = input.handlerInput.requestEnvelope;
        envelope.session.attributes = this.sessionAttributes; // populate saved state
        // ********** INVOKE SKILL ****************
        const responseEnvelope = await this.skill.invoke(envelope, undefined);
        this.sessionAttributes = responseEnvelope.sessionAttributes; // save updated state
        const promptSSML = responseEnvelope.response.outputSpeech.ssml;
        const prompt = promptSSML.replace('<speak>', '').replace('</speak>', '');
        const repromptSSML = responseEnvelope.response.reprompt.outputSpeech.ssml;
        const reprompt = repromptSSML.replace('<speak>', '').replace('</speak>', '');
        const directive = responseEnvelope.response.directives;
        // const cardContent = (responseEnv.response.card as ui.SimpleCard).content;
        return {
            responseEnvelope,
            response: responseEnvelope.response,
            prompt,
            reprompt,
            directive
        };
    }
}
exports.SkillInvoker = SkillInvoker;
